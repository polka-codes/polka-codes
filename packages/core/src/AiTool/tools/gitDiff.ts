// generated by polka.codes
import { z } from 'zod'
import { type FullToolInfoV2, PermissionLevel, type ToolHandler, type ToolInfoV2, ToolResponseType } from '../../tool'
import type { CommandProvider } from '../../tools'

export const toolInfo = {
  name: 'git_diff',
  description:
    'Get the git diff for the current repository. Can be used to get staged changes, unstaged changes, or changes between commits. By default, it returns unstaged changes.',
  parameters: z.object({
    staged: z
      .preprocess((val) => {
        if (typeof val === 'string') {
          const lower = val.toLowerCase()
          if (lower === 'false') return false
          if (lower === 'true') return true
        }
        return val
      }, z.boolean().optional().default(false))
      .describe('Get staged changes instead of unstaged changes.'),
    commitRange: z.string().optional().describe('The commit range to get the diff for (e.g., "main...HEAD").'),
    file: z.string().optional().describe('Get the diff for a specific file.'),
  }),
  permissionLevel: PermissionLevel.Read,
} as const satisfies ToolInfoV2

export const handler: ToolHandler<typeof toolInfo, CommandProvider> = async (provider, args) => {
  if (!provider.executeCommand) {
    return {
      type: ToolResponseType.Error,
      message: 'Not possible to execute command. Abort.',
    }
  }

  const { staged, file, commitRange } = toolInfo.parameters.parse(args)

  const commandParts = ['git', 'diff', '--no-color']
  if (staged) {
    commandParts.push('--staged')
  }
  if (commitRange) {
    commandParts.push(commitRange)
  }
  if (file) {
    commandParts.push('--', file)
  }

  const command = commandParts.join(' ')
  try {
    const result = await provider.executeCommand(command, false)
    if (result.exitCode === 0) {
      if (!result.stdout.trim()) {
        return {
          type: ToolResponseType.Reply,
          message: 'No diff found.',
        }
      }
      return {
        type: ToolResponseType.Reply,
        message: `<diff file="${file ?? 'all'}">\n${result.stdout}\n</diff>`,
      }
    }
    return {
      type: ToolResponseType.Error,
      message: `\`${command}\` exited with code ${result.exitCode}:\n${result.stderr}`,
    }
  } catch (error) {
    return {
      type: ToolResponseType.Error,
      message: error instanceof Error ? error.message : String(error),
    }
  }
}

export const isAvailable = (provider: CommandProvider): boolean => {
  return !!provider.executeCommand
}

export default {
  ...toolInfo,
  handler,
  isAvailable,
} satisfies FullToolInfoV2
