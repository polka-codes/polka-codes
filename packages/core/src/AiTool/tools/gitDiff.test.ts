// generated by polka.codes
import { describe, expect, it, spyOn } from 'bun:test'
import type { CommandProvider } from '../../tools'
import gitDiff from './gitDiff'

describe('gitDiff', () => {
  const mockProvider: CommandProvider = {
    executeCommand: async (command: string) => {
      if (command.includes('-U5')) {
        return {
          exitCode: 0,
          stdout: `@@ -10,3 +10,3 @@
 context line
-old line
+new line`,
          stderr: '',
        }
      }
      if (command.includes('-U5')) {
        return {
          exitCode: 0,
          stdout: `@@ -10,50 +10,50 @@
 many context lines here
-old line
+new line
 many more context lines`,
          stderr: '',
        }
      }
      return { exitCode: 0, stdout: 'default diff output', stderr: '' }
    },
  }

  it('should use default context lines of 5', async () => {
    const spy = spyOn(mockProvider, 'executeCommand')
    await gitDiff.handler(mockProvider, {})

    expect(spy).toHaveBeenCalledWith('git diff --no-color -U5', false)
  })

  it('should use custom context lines when specified', async () => {
    const spy = spyOn(mockProvider, 'executeCommand')
    await gitDiff.handler(mockProvider, { contextLines: '5' })

    expect(spy).toHaveBeenCalledWith('git diff --no-color -U5', false)
  })

  it('should include line numbers when requested', async () => {
    const result = await gitDiff.handler(mockProvider, {
      contextLines: '5',
      includeLineNumbers: 'true',
    })

    expect(result).toMatchSnapshot()
  })

  it('should not include line numbers by default', async () => {
    const result = await gitDiff.handler(mockProvider, { contextLines: '5' })

    expect(result).toMatchSnapshot()
  })

  it('should handle staged changes', async () => {
    const spy = spyOn(mockProvider, 'executeCommand')
    await gitDiff.handler(mockProvider, { staged: 'true' })

    expect(spy).toHaveBeenCalledWith('git diff --no-color -U5 --staged', false)
  })

  it('should handle commit ranges', async () => {
    const spy = spyOn(mockProvider, 'executeCommand')
    await gitDiff.handler(mockProvider, { commitRange: 'main...HEAD' })

    expect(spy).toHaveBeenCalledWith('git diff --no-color -U5 main...HEAD', false)
  })

  it('should handle specific files', async () => {
    const spy = spyOn(mockProvider, 'executeCommand')
    await gitDiff.handler(mockProvider, { file: 'src/test.ts' })

    expect(spy).toHaveBeenCalledWith('git diff --no-color -U5 -- src/test.ts', false)
  })

  it('should combine all options correctly', async () => {
    const spy = spyOn(mockProvider, 'executeCommand')
    await gitDiff.handler(mockProvider, {
      staged: 'true',
      file: 'src/test.ts',
      contextLines: '10',
      includeLineNumbers: 'true',
    })

    expect(spy).toHaveBeenCalledWith('git diff --no-color -U10 --staged -- src/test.ts', false)
  })

  it('should handle no diff found', async () => {
    const emptyProvider: CommandProvider = {
      executeCommand: async () => ({ exitCode: 0, stdout: '', stderr: '' }),
    }

    const result = await gitDiff.handler(emptyProvider, {})
    expect(result).toMatchSnapshot()
  })

  it('should handle command errors', async () => {
    const errorProvider: CommandProvider = {
      executeCommand: async () => ({ exitCode: 1, stdout: '', stderr: 'error message' }),
    }

    const result = await gitDiff.handler(errorProvider, {})
    expect(result).toMatchSnapshot()
  })

  it('should handle missing executeCommand', async () => {
    const result = await gitDiff.handler({} as CommandProvider, {})
    expect(result).toMatchSnapshot()
  })
})
