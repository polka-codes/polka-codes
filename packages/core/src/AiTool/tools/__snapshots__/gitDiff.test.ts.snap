// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`gitDiff should include line numbers when requested 1`] = `
{
  "message": {
    "type": "text",
    "value": 
"<diff file="src/test.ts">
@@ -10,3 +10,3 @@
 context line
-old line [Line 11 removed]
+new line [Line 11]
</diff>"
,
  },
  "type": "Reply",
}
`;

exports[`gitDiff should not include line numbers by default 1`] = `
{
  "message": {
    "type": "text",
    "value": 
"<diff file="src/test.ts">
@@ -10,3 +10,3 @@
 context line
-old line
+new line
</diff>"
,
  },
  "type": "Reply",
}
`;

exports[`gitDiff should handle no diff found 1`] = `
{
  "message": {
    "type": "text",
    "value": "No diff found.",
  },
  "type": "Reply",
}
`;

exports[`gitDiff should handle command errors 1`] = `
{
  "message": {
    "type": "error-text",
    "value": 
"\`git diff --no-color -U5 -- 'src/test.ts'\` exited with code 1:
error message"
,
  },
  "type": "Error",
}
`;

exports[`gitDiff should handle missing executeCommand 1`] = `
{
  "message": {
    "type": "error-text",
    "value": "Not possible to execute command. Abort.",
  },
  "type": "Error",
}
`;
